src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs

```diff
@@ -0,0 +1,237 @@
+                    return false;
+                }
+
+                var methodSymbol = semanticModel.GetSymbolInfo(memberName, cancellationToken).Symbol as IMethodSymbol;
+                if (methodSymbol == null)
+                {
+                    return false;
+                }
+
+                if (memberName.IsRightSideOfDot())
+                {
+                    var typeName = memberName.GetLeftSideOfDot();
+                    return IsPossibleCreationOrConversionMethod(methodSymbol, declaredTypeSymbol, semanticModel, typeName, cancellationToken);
+                }
+
+                return false;
+            }
+
+            private bool IsIntrinsicType(SyntaxNode declarationStatement) =>
+                declarationStatement.IsKind(SyntaxKind.VariableDeclaration)
+                ? ((VariableDeclarationSyntax)declarationStatement).Variables.Single().Initializer.Value.IsAnyLiteralExpression()
+                : false;
+
+            private ExpressionSyntax GetInitializerExpression(EqualsValueClauseSyntax initializer) =>
+                initializer.Value is CheckedExpressionSyntax
+                    ? ((CheckedExpressionSyntax)initializer.Value).Expression
+                    : initializer.Value;
+
+            private bool IsPossibleCreationOrConversionMethod(IMethodSymbol methodSymbol, ITypeSymbol declaredType, SemanticModel semanticModel, ExpressionSyntax typeName, CancellationToken cancellationToken)
+            {
+                if (methodSymbol.ReturnsVoid)
+                {
+                    return false;
+                }
+
+                var typeInInvocation = semanticModel.GetTypeInfo(typeName, cancellationToken).Type;
+
+                return IsPossibleCreationMethod(methodSymbol, declaredType, typeInInvocation)
+                    || IsPossibleConversionMethod(methodSymbol, declaredType, typeInInvocation, semanticModel, cancellationToken);
+            }
+
+            private bool IsPossibleCreationMethod(IMethodSymbol methodSymbol, ITypeSymbol declaredType, ITypeSymbol typeInInvocation)
+            {
+                var isTypeInfoSame = false;
+
+                // Pattern: Method name is a prefix match of one of these well known method names
+                //          and method is a member of type being created.
+                // cases: int.ParseXXX, TextSpan.FromBounds, 
+                if (methodSymbol.Name.StartsWith("Parse", StringComparison.Ordinal)
+                    || methodSymbol.Name.StartsWith("From", StringComparison.Ordinal))
+                {
+                    isTypeInfoSame = typeInInvocation.Equals(declaredType);
+                }
+
+                // Pattern: Method name is an exact match of one of these well known creation methods
+                //          and method is a member of type being created.
+                // cases: node.With, Tuple.Create, State.Generate
+                if (methodSymbol.Name.Equals("With", StringComparison.Ordinal)
```

Comment:
Sorta.

I'm explicitly stating that i want this just for static methods (since the type is apparent in the the type name).

"With" is somewhat special since it'll be an instance method.  But i'm somewhat ok with it since we're codifying "Withers" as a pattern that we use to produce new instances of the same type.

Note those: i'm not sure if "With" actually fits the definition of an apperant type.

if i have:

```c#
var x = blooby.With(...);
```

The type of x is not really apperant at all.  As opposed to:

```c#
var x = Snarfle.SomethingThatReturnsASnarfle();
```

For a type to be apperant, i would expect that i could actually *see* it on the right side somewhere.  in the "With" case, i don't see how that's so.